[FSDP][dtensor] add FSDP2+TP distributed state dict test

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130760
* __->__ #131408
* #132391
* #130239
* #126697

**Test**
`pytest test/distributed/_composable/fsdp/test_fully_shard_training.py`
`pytest test/distributed/_composable/fsdp/test_fully_shard_state_dict.py`
`pytest test/distributed/checkpoint/fsdp/test_fsdp_dsd.py`
`pytest test/distributed/_composable/fsdp/test_fully_shard_init.py`

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @LucasLLC @MeetVadakkanchery @mhorowitz

XilunWu committed:
[WIP][dtensor] add FSDP2+TP distributed state dict test

[ghstack-poisoned]

XilunWu committed:
Update on "[WIP][dtensor] add FSDP2+TP distributed state dict test"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery mhorowitz

[ghstack-poisoned]

XilunWu committed:
Update on "[WIP][dtensor] add FSDP2+TP distributed state dict test"


**Test**
`pytest test/distributed/_composable/fsdp/test_fully_shard_training.py`
`pytest test/distributed/_composable/fsdp/test_fully_shard_state_dict.py`
`pytest test/distributed/checkpoint/fsdp/test_fsdp_dsd.py`
`pytest test/distributed/_composable/fsdp/test_fully_shard_init.py`

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery mhorowitz

[ghstack-poisoned]

XilunWu committed:
Update on "[WIP][dtensor] add FSDP2+TP distributed state dict test"


**Test**
`pytest test/distributed/_composable/fsdp/test_fully_shard_training.py`
`pytest test/distributed/_composable/fsdp/test_fully_shard_state_dict.py`
`pytest test/distributed/checkpoint/fsdp/test_fsdp_dsd.py`
`pytest test/distributed/_composable/fsdp/test_fully_shard_init.py`

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery mhorowitz

[ghstack-poisoned]

XilunWu committed:
Update on "[FSDP][dtensor] add FSDP2+TP distributed state dict test"


**Test**
`pytest test/distributed/_composable/fsdp/test_fully_shard_training.py`
`pytest test/distributed/_composable/fsdp/test_fully_shard_state_dict.py`
`pytest test/distributed/checkpoint/fsdp/test_fsdp_dsd.py`
`pytest test/distributed/_composable/fsdp/test_fully_shard_init.py`

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery mhorowitz

[ghstack-poisoned]

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60606116).

XilunWu committed:
Update on "[FSDP][dtensor] add FSDP2+TP distributed state dict test"


**Test**
`pytest test/distributed/_composable/fsdp/test_fully_shard_training.py`
`pytest test/distributed/_composable/fsdp/test_fully_shard_state_dict.py`
`pytest test/distributed/checkpoint/fsdp/test_fsdp_dsd.py`
`pytest test/distributed/_composable/fsdp/test_fully_shard_init.py`

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery mhorowitz

Differential Revision: [D60606116](https://our.internmc.facebook.com/intern/diff/D60606116)

[ghstack-poisoned]

fegin commented on test/distributed/checkpoint/fsdp/test_fsdp_dsd.py:
```diff
@@ -182,6 +187,392 @@ def _test_save_with_fsdp1_and_load_with_fsdp2(self, state_dict_type: StateDictTy
+            base_model = nn.Sequential(MLP(mlp_dim), MLP(mlp_dim), MLP(mlp_dim))
+            return base_model
+
+        state_dict_type_list = [
+            StateDictType.FULL_STATE_DICT,
+            StateDictType.SHARDED_STATE_DICT,
+        ]
+
+        for state_dict_type in state_dict_type_list:
```
This seems to be redundant to me. Can we save both the FSDP1 state_dict and baseline state_dict? Basically, when calling `dcp.save()`, you can pass something like
```
state_dict = {
   "msd1": baseline_state_dict,
   "osd1": baseline_optimizer_state_dict,
   "msd2": fsdp1_state_dict,
   "osd2": fsdp1_optimizer_state_dict,
}
```
This can also simplify the test code.

fegin commented on test/distributed/checkpoint/fsdp/test_fsdp_dsd.py:
```diff
@@ -182,6 +187,392 @@ def _test_save_with_fsdp1_and_load_with_fsdp2(self, state_dict_type: StateDictTy
+        )
+        dp_mesh, tp_mesh = global_mesh_2d["dp"], global_mesh_2d["tp"]
+
+        for save_full_state_dict in [True, False]:
```
ditto, since we have the baseline osd/msd, we don't need to save the TP state_dict.

fegin commented on test/distributed/checkpoint/fsdp/test_fsdp_dsd.py:
```diff
@@ -182,6 +187,392 @@ def _test_save_with_fsdp1_and_load_with_fsdp2(self, state_dict_type: StateDictTy
+        )
+        dp_mesh, tp_mesh = global_mesh_2d["dp"], global_mesh_2d["tp"]
+
+        for save_full_state_dict in [True, False]:
```
Unlike the previous tests, we may want to preserve both tests because we are testing `full_state_dict` for FSDP2 + TP, which involve strided sharding.

fegin reviewed (commented):
LGTM in general, but we can simplify the test code.

XilunWu commented on test/distributed/checkpoint/fsdp/test_fsdp_dsd.py:
```diff
@@ -182,6 +187,392 @@ def _test_save_with_fsdp1_and_load_with_fsdp2(self, state_dict_type: StateDictTy
+                options=StateDictOptions(full_state_dict=True, cpu_offload=True),
+            )
+
+            with FSDP.state_dict_type(fsdp1_model, state_dict_type):
+                fsdp1_state_dict = {
+                    "model": fsdp1_model.state_dict(),
+                    "optim": FSDP.sharded_optim_state_dict(fsdp1_model, fsdp1_optim),
+                }
```
I tried the following and got `NotImplementedError: Only single local shard is supported.` error:
```
fsdp1_msd = get_model_state_dict(
    fsdp1_model,
    options=StateDictOptions(full_state_dict=save_full_state_dict), 
)
```

trace:
```
File "/data/users/xilunwu/oss/pytorch/test/distributed/checkpoint/fsdp/test_fsdp_dsd.py", line 238, in test_save_with_fsdp1_and_load_with_fsdp2_tp
    fsdp1_msd = get_model_state_dict(
  File "/data/users/xilunwu/oss/pytorch/torch/distributed/checkpoint/state_dict.py", line 991, in get_model_state_dict
    model_state_dict = _get_model_state_dict(model, info)
  File "/data/users/xilunwu/oss/pytorch/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "/data/users/xilunwu/oss/pytorch/torch/distributed/checkpoint/state_dict.py", line 523, in _get_model_state_dict
    if torch.is_tensor(p) and p.is_meta:
  File "/home/xilunwu/local/miniconda3/envs/torchtitan/lib/python3.10/site-packages/typing_extensions.py", line 2636, in wrapper
    return arg(*args, **kwargs)
  File "/data/users/xilunwu/oss/pytorch/torch/distributed/_shard/sharded_tensor/api.py", line 1168, in __torch_function__
    return dispatch(st_instance, func)
  File "/data/users/xilunwu/oss/pytorch/torch/distributed/_shard/sharded_tensor/api.py", line 1149, in dispatch
    return _SHARDED_OPS[func](types, args, kwargs, st._process_group)
  File "/data/users/xilunwu/oss/pytorch/torch/distributed/_shard/op_registry_utils.py", line 38, in wrapper
    return wrapped_func(types, args, kwargs, process_group)
  File "/data/users/xilunwu/oss/pytorch/torch/distributed/_shard/sharded_tensor/_ops/tensor_ops.py", line 59, in st_is_meta
    return args[0].local_tensor().is_meta
  File "/data/users/xilunwu/oss/pytorch/torch/distributed/_shard/sharded_tensor/api.py", line 1131, in local_tensor
    raise NotImplementedError("Only single local shard is supported.")
NotImplementedError: Only single local shard is supported.
```

wz337 replied:
Oh. There is a corner case that ShardedTensor doesn't handle. TLDR is to pass a device_mesh to FSDP1 call so that the state dict would be in DTensor instead.

More context: https://github.com/pytorch/pytorch/issues/132366#issuecomment-2264642034

XilunWu replied:
Thanks @wz337 for the suggestion on using device_mesh argument, and it works!

fegin replied:
What I don't understand is that why @XilunWu didn't encounter such an issue when using FSDP1 native state_dict API. I'll verify if something is wrong with get_state_dict.

XilunWu committed:
Update on "[FSDP][dtensor] add FSDP2+TP distributed state dict test"


**Test**
`pytest test/distributed/_composable/fsdp/test_fully_shard_training.py`
`pytest test/distributed/_composable/fsdp/test_fully_shard_state_dict.py`
`pytest test/distributed/checkpoint/fsdp/test_fsdp_dsd.py`
`pytest test/distributed/_composable/fsdp/test_fully_shard_init.py`

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery mhorowitz

Differential Revision: [D60606116](https://our.internmc.facebook.com/intern/diff/D60606116)

[ghstack-poisoned]

XilunWu committed:
Update on "[FSDP][dtensor] add FSDP2+TP distributed state dict test"


**Test**
`pytest test/distributed/_composable/fsdp/test_fully_shard_training.py`
`pytest test/distributed/_composable/fsdp/test_fully_shard_state_dict.py`
`pytest test/distributed/checkpoint/fsdp/test_fsdp_dsd.py`
`pytest test/distributed/_composable/fsdp/test_fully_shard_init.py`

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery mhorowitz

Differential Revision: [D60606116](https://our.internmc.facebook.com/intern/diff/D60606116)

[ghstack-poisoned]

fegin reviewed (approved):
LGTM

XilunWu committed:
Update on "[FSDP][dtensor] add FSDP2+TP distributed state dict test"


**Test**
`pytest test/distributed/_composable/fsdp/test_fully_shard_training.py`
`pytest test/distributed/_composable/fsdp/test_fully_shard_state_dict.py`
`pytest test/distributed/checkpoint/fsdp/test_fsdp_dsd.py`
`pytest test/distributed/_composable/fsdp/test_fully_shard_init.py`

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery mhorowitz

Differential Revision: [D60606116](https://our.internmc.facebook.com/intern/diff/D60606116)

[ghstack-poisoned]

XilunWu commented:
@pytorchbot merge

XilunWu commented:
@pytorchbot merge

XilunWu commented:
@pytorchbot merge
